### PR TITLE
Serialize stub for dynamic profiles

### LINKED ISSUE #9962: Sometimes, saving a setting in the UI doesn't actually apply
The minimal repro:

* Install powershell core, at least one of them
* Close the Terminal.
* Delete all the settings.json files.
* Open the Terminal, go to the SUI, go to the "Appearance" of the Powershell Core.
* Change scrollbar state to hidden.
* Hit Save.

At this point, the SUI will reload. The scrollbar state didn't apply, and now you've got a mysterious "Default" profile. Also, the `setting.json` is now corrupted. It's now got 

```
{},
{
  "scrollbarState": "hidden"
}
```

slammed in the profiles list, which is garbage data.

* [ ] Figure out why the setting is blasting off to space.
* [ ] Add a mitigation to ignore profile snippets without a GUID

/cc @carlos-zamora. 

Now, I'm gonna close all the threads we think are a dupe of this one.

### PR DESCRIPTION
**Pull request by carlos-zamora**
#9962 was caused by a serialization bug. _Technically_, `ToJson` works
as intended: if the current layer has any values set, write them out to
the json. However, on first load, the dynamic profile `Profile` objects
are actually empty (because they inherit from base layer, then the
dynamic profile generator). This means that `ToJson` writes the dynamic
profiles as empty objects `{}`. Then, on reload, we see that the dynamic
profiles aren't in the JSON, and we write them again.

To get around this issue, we added a simple check to `Profile::ToJson`:
if we have a source, make sure we write out the name, guid, hidden, and
source. This is intended to align with `Profile::GenerateStub`.

Closes #9962

### CODE REVIEW

**Comment by DHowett** [comment]
Great! Thanks so much!

qq: why force it during serialization instead of making sure the values are populated when the layer is created? On deserialization, that is.

or: should we just _always_ write out name, guid, hidden, source?

**Comment by carlos-zamora** [comment]
Had an offline discussion:
- Technically the only important ones to carry over are `guid` and `source`
- `name` is just nice to have because it helps the user orient themselves
- **Idea**: generate `guid`, `source`, and (as comment) `name` (make sure that `GenerateStub` does the same though)

> qq: why force it during serialization instead of making sure the values are populated when the layer is created? On deserialization, that is.

No preference between the two. Figured altering ToJson would be safer but the other would work too. I guess it would mess with finding the source profile object for those 4 properties?

**Review by lhecker — APPROVED** [review]
I can confirm that this fixes the bug of empty (`{}`) profiles being created, as well as the SUI save button not working during the first launch.

I personally believe that we should put this logic into `Profile::ToJson`, as this model class controls how it's serialized/deserialized and "should" probably also control which fields are required in a serialization.
The alternative would be to put this change into `CascadiaSettings::_ApplyDefaultsFromUserSettings` which I believe might be a less general solution.

**Review by zadjii-msft — APPROVED** [review]
I'm gonna approve for now because this is such a high hitter, but we really should have a test for this

File: src/cascadia/TerminalSettingsModel/Profile.cpp [hunk 1]
```diff
@@ -496,11 +496,16 @@ Json::Value Profile::ToJson() const
     // Initialize the json with the appearance settings
     Json::Value json{ winrt::get_self<implementation::AppearanceConfig>(_DefaultAppearance)->ToJson() };
 
+    // GH #9962:
+    //   If the settings.json was missing, when we load the dynamic profiles, they are completely empty.
+    //   This caused us to serialize empty profiles "{}" on accident.
+    bool writeBasicSettings{ !Source().empty() };
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Profile.cpp:502]:
```suggestion
    const bool writeBasicSettings{ !Source().empty() };
```

**Review by DHowett — COMMENTED** [review]
Just a nit. This is a fix that will work, but it is a bit adhoc. To @lhecker's point about things that should be invariant during serialization: _these things should be invariant during construction!_ The object should not be able to be in a bad state, at least for GUID. Name is neither here nor there. :smile:

**Comment by lhecker** [comment]
> these things should be invariant during construction

@DHowett As you might've already heard before the underlying issue is that during the first start we generate dynamic profiles. Due to that you might have two `OriginTag::InBox` profiles (cmd, powershell) and two `OriginTag::Generated` ones (pwsh7, wsl). During this initial load all of them get child profiles in [`CascadiaSettings::_ApplyDefaultsFromUserSettings`](https://github.com/microsoft/terminal/blob/c95ed72aab3b3bfefda9bec94a817726d1a4b43a/src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp#L1016-L1051). At this point we got a problem already: All 4 child profiles have no GUID and no name! The invariant is already violated. So question No. 1:
👉 Would you suggest that `CreateChild()` copies over the GUID, name, etc. to uphold this "invariant"?

This "issue" of having no GUID (and so on) is fixed for the two inbox profiles [right here](https://github.com/microsoft/terminal/blob/c95ed72aab3b3bfefda9bec94a817726d1a4b43a/src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp#L889-L895). Since we load the builtin default-settings.json during load the 2 inbox profiles are stored in `_userSettings` and that's the two profiles which `_GetProfilesJsonObject` returns. As such those two inbox profiles are layered in `_LayerOrCreateProfile`, where they're assigned the 4 essential properties (GUID, ...).

💣 The dynamic profiles don't get this treatment since they're not part of the `_userSettings` during the first start of our app. Only on the second start they'll be stored in `_userSettings`.

**Pull request merged**